File: app/src/main/java/com/tmdb/codechallenge/details/ui/DetailsFragment.java
package com.tmdb.codechallenge.details.ui;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.tmdb.codechallenge.R;
import com.tmdb.codechallenge.api.model.Genre;
import com.tmdb.codechallenge.api.model.Movie;
import com.tmdb.codechallenge.api.model.MovieImage;
import com.tmdb.codechallenge.details.presentation.DetailsViewModel;
import com.tmdb.codechallenge.details.presentation.DetailsViewModelFactory;
import com.tmdb.codechallenge.util.MovieImageUrlBuilder;

import java.util.List;
import java.util.stream.Collectors;

public class DetailsFragment extends Fragment {
    private static final String MOVIE = "movie";

    private Movie movie;
    private DetailsViewModel detailsViewModel;

    private RecyclerView moviePostersCarousel;

    public static DetailsFragment newInstance(Movie movie) {
        DetailsFragment fragment = new DetailsFragment();

        Bundle fragArgs = new Bundle();
        fragArgs.putParcelable(MOVIE, movie);

        fragment.setArguments(fragArgs);

        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle args = getArguments();
        assert args != null;
        movie = (Movie) args.get(MOVIE);
        detailsViewModel = new ViewModelProvider(requireActivity(), new DetailsViewModelFactory(movie)).get(DetailsViewModel.class);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        observeViewModel();
    }

    @Nullable
    @Override
    public View onCreateView(
            @NonNull LayoutInflater inflater,
            @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState
    ) {
        return inflater.inflate(R.layout.fragment_details, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        ImageView imageViewMovieBanner = view.findViewById(R.id.imageViewMovieBanner);

        Glide.with(this)
                .load(new MovieImageUrlBuilder().buildBackdropUrl(movie.getBackdropPath()))
                .into(imageViewMovieBanner);

        TextView textViewMovieSynopses = view.findViewById(R.id.textViewMovieSynopses);
        textViewMovieSynopses.setText(movie.getOverview());

        TextView textViewMovieGenres = view.findViewById(R.id.textViewMovieGenres);
        String genres = genresListToString(movie.getGenres());
        textViewMovieGenres.setText(genres);

        TextView textViewMovieReleaseDate = view.findViewById(R.id.textViewMovieReleaseDate);
        textViewMovieReleaseDate.setText(movie.getReleaseDate());

        moviePostersCarousel = view.findViewById(R.id.moviePostersCarousel);
    }

    private void observeViewModel() {
        detailsViewModel.getImages().observe(this, this::setupCarousel);
        detailsViewModel.retrieveImages();
    }

    private void setupCarousel(List<MovieImage> images) {
        DetailsAdapter detailsAdapter = new DetailsAdapter(images);
        LinearLayoutManager layoutManager = new LinearLayoutManager(requireContext());
        layoutManager.setOrientation(LinearLayoutManager.HORIZONTAL);

        moviePostersCarousel.setLayoutManager(layoutManager);
        moviePostersCarousel.setAdapter(detailsAdapter);
    }

    private String genresListToString(@Nullable List<Genre> genres) {
        assert genres != null;
        return genres.stream()
                .map(Genre::getName)
                .collect(Collectors.joining(", "));
    }
}

File: app/src/main/java/com/tmdb/codechallenge/details/ui/DetailsFragment.kt
package com.tmdb.codechallenge.details.ui

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.bumptech.glide.Glide
import com.tmdb.codechallenge.R
import com.tmdb.codechallenge.api.model.Movie
import com.tmdb.codechallenge.api.model.MovieImage
import com.tmdb.codechallenge.details.presentation.DetailsViewModel
import com.tmdb.codechallenge.details.presentation.DetailsViewModelFactory
import com.tmdb.codechallenge.util.MovieImageUrlBuilder
import kotlinx.android.synthetic.main.fragment_details.*

class DetailsFragment : Fragment() {

    private lateinit var movie: Movie

    private lateinit var detailsViewModel: DetailsViewModel

    private var detailsAdapter: DetailsAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val args = arguments!!
        movie = args[MOVIE] as Movie
        detailsViewModel = ViewModelProvider(requireActivity(), DetailsViewModelFactory(movie)).get(DetailsViewModel::class.java)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        observeViewModel()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_details, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        Glide.with(this)
                .load(MovieImageUrlBuilder().buildBackdropUrl(movie.backdropPath))
                .into(imageViewMovieBanner)

        textViewMovieSynopses.text = movie.overview
        textViewMovieGenres.text = movie.genres?.joinToString(separator = ", ") { it.name }
        textViewMovieReleaseDate.text = movie.releaseDate
    }

    private fun observeViewModel() {
        detailsViewModel.images.observe(viewLifecycleOwner, Observer {
            val images = it ?: return@Observer
            setupCarousel(images)
        })

        detailsViewModel.getImages()
    }

    private fun setupCarousel(images: List<MovieImage>) {
        detailsAdapter = DetailsAdapter(images)
        val layoutManager = LinearLayoutManager(context)
        layoutManager.orientation = LinearLayoutManager.HORIZONTAL
        moviePostersCarousel.layoutManager = layoutManager
        moviePostersCarousel.adapter = detailsAdapter
    }

    companion object {
        private const val MOVIE = "movie"

        fun newInstance(movie: Movie): DetailsFragment {
            val fragArgs = Bundle()
            fragArgs.putParcelable(MOVIE, movie)

            val fragment = DetailsFragment()
            fragment.arguments = fragArgs

            return fragment
        }
    }
}
